File: src/emu/cpu/m6502/m6502make.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdarg.h>
#include <ctype.h>
#include "osdcomm.h"

#include <string>
#include <vector>

using namespace std;

enum { STATES = 0x101 };

const char *snames[STATES-0x100] = {
	"STATE_RESET"
};

struct opcode {
	string name;
	vector<string> instructions;
};

vector<opcode> opcodes;

struct table_t {
	string opcodes[STATES];
};

table_t table;

string device_name;

static void load_opcodes(const char *fname)
{
	char buf[4096];
	FILE *f;

	sprintf(buf, "Error opening %s for reading\n", fname);
	f = fopen(fname, "r");
	if(!f) {
		perror(buf);
		exit(1);
	}

	while(fgets(buf, sizeof(buf), f)) {
		char *p = buf;
		if(p[0] == '#' || p[0] == '\r' || p[0] == '\n' || p[0] == 0)
			continue;

		for(char *q = p; *q; q++)
			if(*q == '\r' || *q == '\n') {
				*q = 0;
				break;
			}

		if(p[0] != ' ' && p[0] != '\t') {
			opcodes.resize(opcodes.size()+1);
			opcodes.back().name = p;
		} else
			opcodes.back().instructions.push_back(p);
	}
	fclose(f);
}

static void load_disp(const char *fname)
{
	char buf[4096];
	FILE *f;

	sprintf(buf, "Error opening %s for reading\n", fname);
	f = fopen(fname, "r");
	if(!f) {
		perror(buf);
		exit(1);
	}

	int state = 0;

	while(fgets(buf, sizeof(buf), f)) {
		char *p = buf;
		if(p[0] == '#' || p[0] == '\r' || p[0] == '\n' || p[0] == 0)
			continue;

		while(state < STATES) {
			while(*p && (*p == '\n' || *p == '\r' || *p == ' ' || *p == '\t'))
				p++;
			if(!*p)
				break;

			char *q = p;
			while(*p != '\n' && *p != '\r' && *p != ' ' && *p != '\t')
				p++;
			table.opcodes[state++] = string(q, p);
		}
	}
	fclose(f);
}

enum { NONE, EAT_ALL, MEMORY };

static int identify_line_type(string inst)
{
	if(inst.find("eat-all-cycles") != string::npos)
		return EAT_ALL;
	if(inst.find("read") != string::npos ||
	   inst.find("write") != string::npos ||
	   inst.find("prefetch(") != string::npos ||
	   inst.find("prefetch_noirq(") != string::npos)
		return MEMORY;
	return NONE;
}

static void save_opcodes(FILE *f)
{
	for(unsigned int i=0; i != opcodes.size(); i++) {
		int substate;
		opcode &o = opcodes[i];

		fprintf(f, "void %s::%s_full()\n", device_name.c_str(), o.name.c_str());
		fprintf(f, "{\n");
		substate = 1;
		for(unsigned int j=0; j != o.instructions.size(); j++) {
			string inst = o.instructions[j];
			int type = identify_line_type(inst);
			if(type == EAT_ALL) {
				fprintf(f, "\ticount=0; inst_substate = %d; return;", substate);
				substate++;
			} else {
				if(type == MEMORY)
					fprintf(f, "\tif(icount == 0) { inst_substate = %d; return; }\n", substate);
				fprintf(f, "%s\n", inst.c_str());
				if(type == MEMORY) {
					fprintf(f, "\ticount--;\n");
					substate++;
				}
			}
		}
		fprintf(f, "}\n");

		fprintf(f, "void %s::%s_partial()\n", device_name.c_str(), o.name.c_str());
		fprintf(f, "{\n");
		fprintf(f, "switch(inst_substate) {\n");
		fprintf(f, "case 0:\n");
		substate = 1;
		for(unsigned int j=0; j != o.instructions.size(); j++) {
			string inst = o.instructions[j];
			int type = identify_line_type(inst);
			if(type == EAT_ALL) {
				fprintf(f, "\ticount=0; inst_substate = %d; return;", substate);
				fprintf(f, "case %d:;\n", substate);
				substate++;
			} else {
				if(type == MEMORY) {
					fprintf(f, "\tif(icount == 0) { inst_substate = %d; return; }\n", substate);
					fprintf(f, "case %d:\n", substate);
				}
				fprintf(f, "%s\n", inst.c_str());
				if(type == MEMORY) {
					fprintf(f, "\ticount--;\n");
					substate++;
				}
			}
		}
		fprintf(f, "}\n");
		fprintf(f, "\tinst_substate = 0;\n");
		fprintf(f, "}\n");
		fprintf(f, "\n");
	}
	fprintf(f, "\n");
}

static void save_tables(FILE *f)
{
	fprintf(f, "void %s::do_exec_full()\n", device_name.c_str());
	fprintf(f, "{\n");
	fprintf(f, "\tswitch(inst_state) {\n");
	for(int j=0; j<STATES; j++) {
		if(table.opcodes[j] != ".") {
			if(j < 0x100)
				fprintf(f, "\tcase 0x%02x: %s_full(); break;\n", j, table.opcodes[j].c_str());
			else
				fprintf(f, "\tcase %s: %s_full(); break;\n", snames[j-0x100], table.opcodes[j].c_str());
		}
	}
	fprintf(f, "\t}\n");
	fprintf(f, "}\n");
	fprintf(f, "void %s::do_exec_partial()\n", device_name.c_str());
	fprintf(f, "{\n");
	fprintf(f, "\tswitch(inst_state) {\n");
	for(int j=0; j<STATES; j++) {
		if(table.opcodes[j] != ".") {
			if(j < 0x100)
				fprintf(f, "\tcase 0x%02x: %s_partial(); break;\n", j, table.opcodes[j].c_str());
			else
				fprintf(f, "\tcase %s: %s_partial(); break;\n", snames[j-0x100], table.opcodes[j].c_str());
		}
	}
	fprintf(f, "\t}\n");
	fprintf(f, "}\n");
	fprintf(f, "const %s::disasm_entry %s::disasm_entries[0x100] = {\n", device_name.c_str(), device_name.c_str());
	for(int j=0; j<0x100; j++)
		if(table.opcodes[j] != ".") {
			string opcode = table.opcodes[j];
			string opc, fullopc, mode;
			string::iterator k, ke;
			for(k = opcode.begin(); k != opcode.end() && *k != '_'; k++);
			for(ke = opcode.end(); ke != opcode.begin() && ke[-1] != '_'; ke--);
			assert(k != opcode.end());
			opc = string(opcode.begin(), k);
			fullopc = string(opcode.begin(), ke-1);
			mode = string(ke, opcode.end());
			
			bool step_over = opc == "jsr" || opc == "bsr";
			bool step_out = opc == "rts" || opc == "rti" || opc == "rtn";
			bool per_bit = opc == "bbr" || opc == "bbs" || opc == "rmb" || opc == "smb";
			fprintf(f, "\t{ \"%s\", DASM_%s, %s, %s },\n",
					opc.c_str(), mode.c_str(), step_over ? "DASMFLAG_STEP_OVER" : step_out ? "DASMFLAG_STEP_OUT" : "0",
					per_bit ? "true" : "false");
		} else
			fprintf(f, "\t{ \"???\", DASM_imp, 0, false },\n");
	fprintf(f, "};\n");
}

static void save(const char *fname)
{
	char buf[4096];
	FILE *f;

	sprintf(buf, "Error opening %s for writing\n", fname);
	f = fopen(fname, "w");
	if(!f) {
		perror(buf);
		exit(1);
	}

	save_opcodes(f);
	save_tables(f);

	fclose(f);
}

int main(int argc, char *argv[])
{
	if(argc != 5) {
		fprintf(stderr, "Usage:\n%s device_name {opc.lst|-} disp.lst device.inc\n", argv[0]);
		exit(1);
	}

	device_name = argv[1];
	if(strcmp(argv[2], "-"))
	   load_opcodes(argv[2]);
	load_disp(argv[3]);
	save(argv[4]);


	return 0;
}

File: src/emu/cpu/m6502/m6502make.py
#!/usr/bin/python

USAGE = """
Usage:
%s device_name {opc.lst|-} disp.lst device.inc
"""
import sys
import logging

MAX_STATES = 0

def load_opcodes(fname):
    """Load opcodes from .lst file"""
    opcodes = []
    logging.info("load_opcodes: %s", fname)
    try:
        f = open(fname, "rU")
    except Exception:
        err = sys.exc_info()[1]
        logging.error("cannot read opcodes file %s [%s]", fname, err)
        sys.exit(1)

    for line in f:
        if line.startswith("#"): continue
        line = line.rstrip()
        if not line: continue
        if line.startswith(" ") or line.startswith("\t"):
            # append instruction to last opcode
            opcodes[-1][1].append(line)
        else:
            # add new opcode
            opcodes.append((line, []))
    return opcodes


def load_disp(fname):
    logging.info("load_disp: %s", fname)
    states = []
    try:
        f = open(fname, "rU")
    except Exception:
        err = sys.exc_info()[1]
        logging.error("cannot read display file %s [%s]", fname, err)
        sys.exit(1)
    for line in f:
        if line.startswith("#"): continue
        line = line.strip()
        if not line: continue
        tokens = line.split()
        states += tokens
    return states

def emit(f, text):
    """write string to file"""
    print >>f, text,

FULL_PROLOG="""\
void %(device)s::%(opcode)s_full()
{
"""

FULL_EPILOG="""\
}
"""

FULL_EAT_ALL="""\
\ticount=0; inst_substate = %(substate)s; return;
"""

FULL_MEMORY="""\
\tif(icount == 0) { inst_substate = %(substate)s; return; }
%(ins)s
\ticount--;
"""

FULL_NONE="""\
%(ins)s
"""

PARTIAL_PROLOG="""\
void %(device)s::%(opcode)s_partial()
{
switch(inst_substate) {
case 0:
"""

PARTIAL_EPILOG="""\
}
\tinst_substate = 0;
}

"""

PARTIAL_EAT_ALL="""\
\ticount=0; inst_substate = %(substate)s; return;
case %(substate)s:;
"""

PARTIAL_MEMORY="""\
\tif(icount == 0) { inst_substate = %(substate)s; return; }
case %(substate)s:
%(ins)s
\ticount--;
"""

PARTIAL_NONE="""\
%(ins)s
"""
def identify_line_type(ins):
    if "eat-all-cycles" in ins: return "EAT"
    for s in ["read", "write", "prefetch(", "prefetch_noirq("]:
        if s in ins:
            return "MEMORY"
    return "NONE"


def save_opcodes(f, device, opcodes):
    for name, instructions in opcodes:
        d = { "device": device,
              "opcode": name,
              }

        emit(f, FULL_PROLOG % d)
        substate = 1
        for ins in instructions:
            d["substate"] = str(substate)
            d["ins"] =  ins
            line_type = identify_line_type(ins)
            if line_type == "EAT":
                emit(f, FULL_EAT_ALL % d)
                substate += 1
            elif line_type == "MEMORY":
                emit(f, FULL_MEMORY % d)
                substate += 1
            else:
                emit(f, FULL_NONE %d)
        emit(f, FULL_EPILOG % d)

        emit(f, PARTIAL_PROLOG % d)
        substate = 1
        for ins in instructions:
            d["substate"] = str(substate)
            d["ins"] =  ins
            line_type = identify_line_type(ins)
            if line_type == "EAT":
                emit(f, PARTIAL_EAT_ALL % d)
                substate += 1
            elif line_type == "MEMORY":
                emit(f, PARTIAL_MEMORY % d)
                substate += 1
            else:
                emit(f, PARTIAL_NONE %d)
        emit(f, PARTIAL_EPILOG % d)


DO_EXEC_FULL_PROLOG="""\
void %(device)s::do_exec_full()
{
\tswitch(inst_state) {
"""

DO_EXEC_FULL_EPILOG="""\
\t}
}
"""

DO_EXEC_PARTIAL_PROLOG="""\
void %(device)s::do_exec_partial()
{
\tswitch(inst_state) {
"""

DO_EXEC_PARTIAL_EPILOG="""\
\t}
}
"""

DISASM_PROLOG="""\
const %(device)s::disasm_entry %(device)s::disasm_entries[0x%(disasm_count)x] = {
"""

DISASM_EPILOG="""\
};
"""

def save_tables(f, device, states):
    total_states = len(states)

    d = { "device": device,
          "disasm_count": total_states-1
          }
    
    
    emit(f, DO_EXEC_FULL_PROLOG % d)
    for n, state in enumerate(states):
        if state == ".": continue
        if n < total_states - 1:
            emit(f, "\tcase 0x%02x: %s_full(); break;\n" % (n, state))
        else:
            emit(f, "\tcase %s: %s_full(); break;\n" % ("STATE_RESET", state))
    emit(f, DO_EXEC_FULL_EPILOG % d)

    emit(f, DO_EXEC_PARTIAL_PROLOG % d)
    for n, state in enumerate(states):
        if state == ".": continue
        if n < total_states - 1:
            emit(f, "\tcase 0x%02x: %s_partial(); break;\n" % (n, state))
        else:
            emit(f, "\tcase %s: %s_partial(); break;\n" % ("STATE_RESET", state))
    emit(f, DO_EXEC_PARTIAL_EPILOG % d)

    emit(f, DISASM_PROLOG % d )
    for n, state in enumerate(states):
        if state == ".": continue
        if n == total_states - 1: break
        tokens = state.split("_")
        opc = tokens[0]
        mode = tokens[-1]
        extra = "0"
        if opc in ["jsr", "bsr"]:
            extra =  "DASMFLAG_STEP_OVER"
        elif opc in ["rts", "rti", "rtn"]:
            extra = "DASMFLAG_STEP_OUT"
        emit(f, '\t{ "%s", DASM_%s, %s },\n' % (opc, mode, extra))
    emit(f, DISASM_EPILOG % d)

def save(fname, device, opcodes, states):
    logging.info("saving: %s", fname)
    try:
        f = open(fname, "w")
    except Exception:
        err = sys.exc_info()[1]
        logging.error("cannot write file %s [%s]", fname, err)
        sys.exit(1)
    save_opcodes(f,device, opcodes)
    emit(f, "\n")
    save_tables(f, device, states)
    f.close()


def main(argv):
    debug = False
    logformat=("%(levelname)s:"
               "%(module)s:"
               "%(lineno)d:"
               "%(threadName)s:"
               "%(message)s")
    if debug:
        logging.basicConfig(level=logging.INFO, format=logformat)
    else:
        logging.basicConfig(level=logging.WARNING, format=logformat)


    if len(argv) != 5:
        print USAGE % argv[0]
        return 1

    device_name = argv[1]

    opcodes = []
    if argv[2] !=  "-":
        opcodes = load_opcodes(argv[2])
        logging.info("found %d opcodes", len(opcodes))
    else:
        logging.info("skipping opcode reading")


    states = load_disp(argv[3])
    logging.info("loaded %s states", len(states))

    assert (len(states) & 0xff) == 1
    save(argv[4], device_name, opcodes, states)


# ======================================================================
if __name__ == "__main__":
    sys.exit(main(sys.argv))

